src/powerfit_em/correlators/gpu.py

```diff
@@ -0,0 +1,302 @@
+        target: np.ndarray,
+        template: np.ndarray,
+        rotations: np.ndarray,
+        mask: np.ndarray,
+        queue: cl.CommandQueue,
+        laplace: bool = False,
+    ):
+        """Initialize the GPU correlator.
+
+        Args:
+            target: the target density on which you want to fit a template structure.
+            template: the template structure that you want to fit in the target density,
+                should have been regridded to the same grid as the target density.
+            rotations: array of 3D-rotations, for each rotation all translation the
+                template is both rotated and translated to compute the highest local cross-
+                correlation score.
+            mask: core-weighted mask. See doi:10.3934/biophy.2015.2.73, Figure 1.
+            queue: the OpenCL command queue on which to execute the computations.
+            laplace: if true, a Laplace pre-filter is applied to the target density and
+                template to enhance the sensitivity of the scoring function.
+        """
+
+        self.target: np.ndarray = target / target.max()
+        self.laplace = laplace
+        self.mask = mask
+        self._queue = queue
+
+        self._rotations = transform_rotations(rotations)
+
+        # Precompute the normalization factor for use in the LCC computing kernel
+        self._norm_factor = np.float32((mask != 0).sum())
+        if self._norm_factor == 0:
+            raise ValueError('Zero-filled mask is not allowed.')
+
+        self.gpu_vars, self.gpu_vars_ft = init_gpu_vars(queue, self.target, mask, self.laplace)
+
+        self.lcc = np.zeros(self.target.shape, dtype=np.float32)
+        self.rot = np.zeros(self.target.shape, dtype=np.int32)
+
+        self.set_template(template)
+
+        self.cl_kernels = generate_kernels(queue, self.target)
+        precompute_squared_targets(self.gpu_vars, self.gpu_vars_ft, self.cl_kernels)
+
+    def set_template(self, template: np.ndarray):
+        """Set the template structure that you want to fit in the target density.
+
+        Can be used to try to fit a different template to the same target structure
+        without recomputing the kernels.
+        
+        Args:
+            template: the template structure that you want to fit in the target density,
+                should have been regridded to the same grid as the target density.
+        """
+        if template.shape != self.target.shape:
+            raise ValueError("Shape of template does not match the target.")
+
+        if self.laplace:
+            template = laplace_filter(template, mode='wrap')
+        
+        template = normalize_template(template, self.mask)
+        self.gpu_vars.template = cl.image_from_array(self.gpu_vars.template.context, template.astype(f32))
+
+        # Reset lcc and rot values after (re)setting the template
+        self.lcc[:] = 0.0
+        self.rot[:] = 0
+
+    @property
+    def queue(self) -> cl.CommandQueue:
```

Comment:
What is the reason for having this prop? 

I only see self._queue used here then why not rename _queue to queue or remove the queue prop and use _queue elsewhere in class.

Reply:
Yeah I can just remove the queue prop. It's not a public API so users should not expect to be able to modify class properties without weird side effects.